File: Edge-Detection-OpenCV/edge_detection.py
#!/usr/bin/env python3
"""Sobel and Canny edge detection with reproducible, headless outputs."""

from __future__ import annotations

import argparse
from dataclasses import dataclass
from pathlib import Path

import cv2
import numpy as np


PROJECT_DIR = Path(__file__).resolve().parent
DEFAULT_INPUT = PROJECT_DIR / "assets" / "sample-scene.png"


@dataclass(frozen=True)
class EdgeResult:
    gray: np.ndarray
    blurred: np.ndarray
    sobel_x: np.ndarray
    sobel_y: np.ndarray
    magnitude: np.ndarray
    canny: np.ndarray


def read_bgr(path: Path) -> np.ndarray:
    image = cv2.imread(str(path), cv2.IMREAD_COLOR)
    if image is None:
        raise FileNotFoundError(f"Could not read input image: {path}")
    return image


def normalize_gradient(gradient: np.ndarray) -> np.ndarray:
    return cv2.normalize(
        np.abs(gradient), None, 0, 255, cv2.NORM_MINMAX, dtype=cv2.CV_8U
    )


def detect_edges(
    image_bgr: np.ndarray,
    low_threshold: int = 100,
    high_threshold: int = 200,
    blur_size: int = 5,
) -> EdgeResult:
    if image_bgr.ndim != 3 or image_bgr.shape[2] != 3:
        raise ValueError("image_bgr must be a three-channel BGR image")
    if image_bgr.dtype != np.uint8:
        raise ValueError("image_bgr must use uint8 pixels")
    if not 0 <= low_threshold < high_threshold <= 255:
        raise ValueError("thresholds must satisfy 0 <= low < high <= 255")
    if blur_size < 3 or blur_size % 2 == 0:
        raise ValueError("blur_size must be an odd integer of at least 3")

    gray = cv2.cvtColor(image_bgr, cv2.COLOR_BGR2GRAY)
    blurred = cv2.GaussianBlur(gray, (blur_size, blur_size), 0)
    sobel_x_f = cv2.Sobel(blurred, cv2.CV_32F, 1, 0, ksize=3)
    sobel_y_f = cv2.Sobel(blurred, cv2.CV_32F, 0, 1, ksize=3)
    magnitude_f = cv2.magnitude(sobel_x_f, sobel_y_f)

    return EdgeResult(
        gray=gray,
        blurred=blurred,
        sobel_x=normalize_gradient(sobel_x_f),
        sobel_y=normalize_gradient(sobel_y_f),
        magnitude=normalize_gradient(magnitude_f),
        canny=cv2.Canny(blurred, low_threshold, high_threshold, L2gradient=True),
    )


def make_comparison(image_bgr: np.ndarray, result: EdgeResult) -> np.ndarray:
    panels = [
        image_bgr.copy(),
        cv2.cvtColor(result.magnitude, cv2.COLOR_GRAY2BGR),
        cv2.cvtColor(result.canny, cv2.COLOR_GRAY2BGR),
    ]
    labels = ["Input", "Sobel magnitude", "Canny"]
    for panel, label in zip(panels, labels):
        cv2.rectangle(panel, (0, 0), (panel.shape[1], 42), (0, 0, 0), -1)
        cv2.putText(
            panel,
            label,
            (14, 29),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.72,
            (255, 255, 255),
            2,
            cv2.LINE_AA,
        )
    return cv2.hconcat(panels)


def write_results(
    output_dir: Path, image_bgr: np.ndarray, result: EdgeResult
) -> dict[str, Path]:
    output_dir.mkdir(parents=True, exist_ok=True)
    outputs = {
        "gray": output_dir / "gray.png",
        "sobel_x": output_dir / "sobel-x.png",
        "sobel_y": output_dir / "sobel-y.png",
        "sobel_magnitude": output_dir / "sobel-magnitude.png",
        "canny": output_dir / "canny.png",
        "comparison": output_dir / "edge-comparison.png",
    }
    images = {
        "gray": result.gray,
        "sobel_x": result.sobel_x,
        "sobel_y": result.sobel_y,
        "sobel_magnitude": result.magnitude,
        "canny": result.canny,
        "comparison": make_comparison(image_bgr.copy(), result),
    }
    for name, path in outputs.items():
        if not cv2.imwrite(str(path), images[name]):
            raise RuntimeError(f"Could not write output image: {path}")
    return outputs


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--input", type=Path, default=DEFAULT_INPUT)
    parser.add_argument("--output-dir", type=Path, default=PROJECT_DIR / "outputs")
    parser.add_argument("--low", type=int, default=100)
    parser.add_argument("--high", type=int, default=200)
    parser.add_argument("--blur-size", type=int, default=5)
    parser.add_argument("--display", action="store_true")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    image = read_bgr(args.input)
    result = detect_edges(image, args.low, args.high, args.blur_size)
    outputs = write_results(args.output_dir, image, result)
    print(
        f"size={image.shape[1]}x{image.shape[0]} "
        f"canny_pixels={cv2.countNonZero(result.canny)} "
        f"outputs={len(outputs)}"
    )
    if args.display:
        cv2.imshow("Sobel and Canny edge detection", make_comparison(image, result))
        cv2.waitKey(0)
        cv2.destroyAllWindows()


if __name__ == "__main__":
    main()

File: Edge-Detection-OpenCV/edge_detection.cpp
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

namespace fs = std::filesystem;

struct EdgeResult {
    cv::Mat gray;
    cv::Mat blurred;
    cv::Mat sobelX;
    cv::Mat sobelY;
    cv::Mat magnitude;
    cv::Mat canny;
};

cv::Mat normalizeGradient(const cv::Mat& gradient) {
    cv::Mat absoluteGradient;
    cv::absdiff(gradient, cv::Scalar::all(0), absoluteGradient);
    cv::Mat output;
    cv::normalize(absoluteGradient, output, 0, 255, cv::NORM_MINMAX, CV_8U);
    return output;
}

EdgeResult detectEdges(const cv::Mat& imageBgr, int low, int high, int blurSize) {
    if (imageBgr.empty() || imageBgr.type() != CV_8UC3) {
        throw std::invalid_argument("input must be a non-empty 8-bit BGR image");
    }
    if (low < 0 || low >= high || high > 255) {
        throw std::invalid_argument("thresholds must satisfy 0 <= low < high <= 255");
    }
    if (blurSize < 3 || blurSize % 2 == 0) {
        throw std::invalid_argument("blur size must be an odd integer of at least 3");
    }

    EdgeResult result;
    cv::cvtColor(imageBgr, result.gray, cv::COLOR_BGR2GRAY);
    cv::GaussianBlur(result.gray, result.blurred, cv::Size(blurSize, blurSize), 0);

    cv::Mat sobelX32f;
    cv::Mat sobelY32f;
    cv::Sobel(result.blurred, sobelX32f, CV_32F, 1, 0, 3);
    cv::Sobel(result.blurred, sobelY32f, CV_32F, 0, 1, 3);
    cv::Mat magnitude32f;
    cv::magnitude(sobelX32f, sobelY32f, magnitude32f);

    result.sobelX = normalizeGradient(sobelX32f);
    result.sobelY = normalizeGradient(sobelY32f);
    result.magnitude = normalizeGradient(magnitude32f);
    cv::Canny(result.blurred, result.canny, low, high, 3, true);
    return result;
}

cv::Mat makeComparison(const cv::Mat& imageBgr, const EdgeResult& result) {
    std::vector<cv::Mat> panels;
    panels.push_back(imageBgr.clone());
    cv::Mat magnitudeBgr;
    cv::Mat cannyBgr;
    cv::cvtColor(result.magnitude, magnitudeBgr, cv::COLOR_GRAY2BGR);
    cv::cvtColor(result.canny, cannyBgr, cv::COLOR_GRAY2BGR);
    panels.push_back(magnitudeBgr);
    panels.push_back(cannyBgr);
    const std::vector<std::string> labels{"Input", "Sobel magnitude", "Canny"};
    for (std::size_t i = 0; i < panels.size(); ++i) {
        cv::rectangle(panels[i], cv::Rect(0, 0, panels[i].cols, 42), cv::Scalar::all(0), -1);
        cv::putText(panels[i], labels[i], cv::Point(14, 29), cv::FONT_HERSHEY_SIMPLEX,
                    0.72, cv::Scalar::all(255), 2, cv::LINE_AA);
    }
    cv::Mat comparison;
    cv::hconcat(panels, comparison);
    return comparison;
}

void requireWrite(const fs::path& path, const cv::Mat& image) {
    if (!cv::imwrite(path.string(), image)) {
        throw std::runtime_error("could not write " + path.string());
    }
}

int selfTest() {
    cv::Mat fixture = cv::Mat::zeros(96, 96, CV_8UC3);
    cv::rectangle(fixture, cv::Rect(20, 24, 50, 44), cv::Scalar::all(255), -1);
    const EdgeResult result = detectEdges(fixture, 50, 150, 5);
    const int edgePixels = cv::countNonZero(result.canny);
    if (result.canny.type() != CV_8UC1 || edgePixels < 150 || edgePixels > 250 ||
        cv::countNonZero(result.magnitude) == 0) {
        std::cerr << "self-test failed: edge_pixels=" << edgePixels << '\n';
        return 1;
    }
    std::cout << "self-test passed: edge_pixels=" << edgePixels << '\n';
    return 0;
}

int main(int argc, char** argv) {
    try {
        fs::path input = fs::path(__FILE__).parent_path() / "assets" / "sample-scene.png";
        fs::path outputDir = fs::path(__FILE__).parent_path() / "outputs";
        int low = 100;
        int high = 200;
        int blurSize = 5;
        bool display = false;
        for (int i = 1; i < argc; ++i) {
            const std::string arg = argv[i];
            if (arg == "--self-test") return selfTest();
            if (arg == "--display") display = true;
            else if (arg == "--input" && i + 1 < argc) input = argv[++i];
            else if (arg == "--output-dir" && i + 1 < argc) outputDir = argv[++i];
            else if (arg == "--low" && i + 1 < argc) low = std::stoi(argv[++i]);
            else if (arg == "--high" && i + 1 < argc) high = std::stoi(argv[++i]);
            else if (arg == "--blur-size" && i + 1 < argc) blurSize = std::stoi(argv[++i]);
            else if (arg != "--display") throw std::invalid_argument("unknown or incomplete option: " + arg);
        }

        const cv::Mat image = cv::imread(input.string(), cv::IMREAD_COLOR);
        if (image.empty()) throw std::runtime_error("could not read " + input.string());
        const EdgeResult result = detectEdges(image, low, high, blurSize);
        fs::create_directories(outputDir);
        requireWrite(outputDir / "gray.png", result.gray);
        requireWrite(outputDir / "sobel-x.png", result.sobelX);
        requireWrite(outputDir / "sobel-y.png", result.sobelY);
        requireWrite(outputDir / "sobel-magnitude.png", result.magnitude);
        requireWrite(outputDir / "canny.png", result.canny);
        const cv::Mat comparison = makeComparison(image, result);
        requireWrite(outputDir / "edge-comparison.png", comparison);
        std::cout << "size=" << image.cols << "x" << image.rows
                  << " canny_pixels=" << cv::countNonZero(result.canny)
                  << " outputs=6\n";
        if (display) {
            cv::imshow("Sobel and Canny edge detection", comparison);
            cv::waitKey(0);
        }
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
        return 1;
    }
}
